Fix MediaHelper::embed() to allow links to be added around embedded medi...

gBokiau opened the pull request:
...a

Commit:
Fix MediaHelper::embed() to allow links to be added around embedded media

ndm2 commented:
This will cause all of the [`HtmlHelper::link()`](https://github.com/cakephp/cakephp/blob/2.4.2/lib/Cake/View/Helper/HtmlHelper.php#L333) functionality to be lost, ie no support for array style URLs and confirm messages anymore. Also `MediaHelper::embedAsObject()` needs to be taken care of too.

I'd suggest that a possible fix should keep the original functionality.

Commit:
parse url parameter

gBokiau commented:
Duh, you're right about url. But since no options were passed to HtmlHelper::link() previously either, this does fully restore previous functionality.

Commit:
and fix embedAsObject as well

ndm2 commented on View/Helper/MediaHelper.php:
```diff
@@ -612,5 +608,12 @@ protected function _encodeUrl($url) {
+		if ($url !== null) {
+			$url = parent::url($url);
+		}
+		return sprintf($this->_tags['link'], $url, array(), $html);
```
Passing `array()` causes an array to string conversion notice, and it results in `Array` being added to the string. Instead `null` or `''` should be passed.

ndm2 commented:
You're right, the link options are hard-coded, so the confirm functionality is irrelevant in terms of backwards compatibility.

gBokiau committed:
fix sloppiness (thx @ndm2)

ndm2 commented on View/Helper/MediaHelper.php:
```diff
@@ -612,5 +608,12 @@ protected function _encodeUrl($url) {
 	}
+	
+	private function _link($html, $url) {
+		if ($url !== null) {
```
Is this ever expected to be `null`?

gBokiau committed:
$url cannot be null by design

ndm2 commented on View/Helper/MediaHelper.php:
```diff
@@ -612,5 +608,10 @@ protected function _encodeUrl($url) {
+	
+	private function _link($html, $url) {
+		$url = parent::url($url);
+		return sprintf($this->_tags['link'], $url, null, $html);
```
Not sure if anyone is with me on this, but what about using `HtmlHelper::useTag()` instead? While I wouldn't expect this to break with future CakePHP 2.x updates, it might still be a good idea to use the abstracted tag generation, `MediaHelper::embed()` uses it too, so it would also be more consistent.

ndm2 commented on View/Helper/MediaHelper.php:
```diff
@@ -612,5 +608,10 @@ protected function _encodeUrl($url) {
 		return str_replace($path, $encoded, $url);
 	}
+	
+	private function _link($html, $url) {
```
If you're not already tired of my notes... I think this should be `protected`? Also the method should have a proper doc-block.

ndm2 commented:
Generally this looks like it fixes the problem while keeping BC (AFAICT). Maybe you can also add some tests to avoid future regressions, and squash the commits into a single one?

gBokiau committed:
Use HtmlHelper::useTag instead of sprintf() and add docs to _link()

gBokiau commented:
Figured we might as well use ::useTag() everywhere. Also integrated the configFile functionality. WDYT?

ndm2 commented:
I'm personally pro `useTag()` and exposing configuration, but I think this is a little out of the scope of this PR, as it has nothing to do with the fix anymore. I'd suggest to keep these possible additional changes separately where they could then be discussed.

gBokiau committed:
Basic test for url option in MediaHelper::embed()

ReplicaObscura commented:
With the possible exception of https://github.com/gBokiau/media/commit/a7022342ea2680a696b6a235fdcf25fb2c7a1200, is there a consensus that this is a good way to solve the reported link issue without introducing any new issues or too much added complexity?

It seems pretty straight-forward to me. And I would tend to agree with @ndm2 that the ::useTag() change is an excellent one, but should probably be a separate pull request so we're not adding extra baggage onto this fix.

ndm2 commented:
I'm still checking for possible implications of future `HtmlHelper` changes, but other than that I'd say it looks good, provided that continuing to extend `HtmlHelper` is being seen as the way to go.

gBokiau committed:
Hardcore simplification

I _think_ this does the exact same thing as previously, at least in the
case of images. The rest needs further looking into

gBokiau commented:
So, I _think_ url wasn't doing anything that webroot hadn't already done.
I _think_, except for the resolving paths part, one can just as well rely on native Helper methods for these two functions, if needed
I think _parseAttributes() effectively only added 'autobuffer' to the list of minimized tags (which I moved to the constructor)
I think we should do the url encoding only at the very end, like HtmlHelper does (and rely on its native methods if they already tackle this)

… seems to be working for images.

ndm2 commented:
Now you're going a little too wild :tiger: I'm generally all for simplification :+1: , but I do see logic errors, failing tests and BC breaking changes here.

Please check the tests first. While they are still not complete (not yet covering default attributes, custom webroots, using `$full` for `url()`, etc), the existing ones need to pass!

See also https://github.com/bmcclure/CakePHP-Media-Plugin/issues/49#issuecomment-28238733
